clang/lib/Format/UnwrappedLineParser.cpp

```diff
@@ -491,11 +491,19 @@ void UnwrappedLineParser::calculateBraceTypes(bool ExpectClassBody) {
   SmallVector<StackEntry, 8> LBraceStack;
   assert(Tok->is(tok::l_brace));
   do {
-    // Get next non-comment token.
-    FormatToken *NextTok;
-    do {
-      NextTok = Tokens->getNextToken();
-    } while (NextTok->is(tok::comment));
+    // Get next non-comment, non-preprocessor token.
+    FormatToken *NextTok = Tokens->getNextToken();
+    while (NextTok->is(tok::comment) ||
+           (NextTok->is(tok::hash) && isOnNewLine(*NextTok))) {
+      while (NextTok->is(tok::comment))
+        NextTok = Tokens->getNextToken();
+      while (NextTok->is(tok::hash) && isOnNewLine(*NextTok)) {
+        ScopedMacroState MacroState(*Line, Tokens, NextTok);
+        do {
+          NextTok = Tokens->getNextToken();
+        } while (NextTok->isNot(tok::eof));
+      }
+    }
```

Comment:
```suggestion
    while (NextTok->is(tok::hash)) {
      NextTok = Tokens->getNextToken();
      do {
        NextTok = Tokens->getNextToken();
      } while (NextTok->is(tok::comment) || (NextTok->NewlinesBefore == 0 &&
                                             NextTok->isNot(tok::eof)));
    } while(NextTok->is(tok::hash));
```
And keep the deleted lines (494-498).

I don't think we need to (or should) use `isOnNewLine()` as it might miss PPDirective lines started with block comments.

Reply:
removing isOnNewLine causes test failures containing macro stringizing operators (which also uses `#`)

Reply:
Can you give a valid C/C++ example? As far as I know, you can't have `#` after `{`.

Reply:
Tested on the following case without `isOnNewLine()`:
```
struct Foo {
  Foo() : a(2) {
#define FOO(foo) \
  /* foo */ #foo
  }

  Foo &operator+=(const Foo &) {
    return *this;
  }
};
```
Also works if the comment is deleted.

Reply:
Ah, sorry, I missed the fact that you gave `NewlinesBefore` as the replacement to `isOnNewLine`, yes, that works better, thank you